### ISSUE TITLE
[Libretro] Using CHDs instead of CUE/BIN/GDI/CDI has a noticeable impact

### ISSUE THREAD

**Issue by vanfanel** [flyinghead/flycast#1755]
**Please Note: This form is the minimum required information for submitting bugs.**  
**Removing this form may lead to your issue being closed until it is completed.**

Platform / OS / Hardware:  GNU/Linux

Flycast version:  latest GIT code

Hardware: Intel 12th Gen + Iris XE graphics

**_Description of the Issue_**

Using compressed images (CHD) has a performance impact on this core, causing audio dropouts if ALSA sound backend is used or cracking if ALSATHREAD is used instead.
This is the only core where I see this (PPSSPP or LRPS2 don't show this small problem), hence my report/question.

So my question is: May it be that the core doesn't use threaded disk access like the rest of the CD based cores?

**Comment by flyinghead** [comment]
I doubt using CHDs would cause any performance issue (and thus audio cracking). Any decent machine is *way* faster at reading and decompressing a CHD than original hardware can read a GD-ROM.
I don't think anybody has ever reported this issue with CHDs, particularly on a modern desktop machine.
Are you using an insanely small audio buffer size/latency setting?

CHDs are an issue for arcade games because the entire disk is read, decompressed and decrypted into memory before the game starts. I changed this so that the CHD is loaded by chunks on demand (dev branch). But this is only arcade games.

**Comment by vanfanel** [comment]
@flyinghead I'm using 32ms audio latency, which isn't insanely small at all and doesn't seem to cause any problems at all with other 3D cores.
Of course increasing the audio buffer size will "mask" this issue, but it will also increase the audio latency when compared to the real machine, which I know very well (as a player! :D).
 
About other latency related settings, well, runahead is of course disabled, and aside of that I haven't set anything special.

**Comment by flyinghead** [comment]
Which game(s) do you have issues with?
Can you also indicate when does the audio dropout/crackle happen?

**Comment by vanfanel** [comment]
> Which game(s) do you have issues with? Can you also indicate when does the audio dropout/cracking happen?

Yes. SoulCalibur: Just as you start the game, when the "Naaaamcowwww" voice plays along with a strong winding sound, you can hear the crackling if you use ALSATHREAD or the dropout if you use ALSA.

**Comment by flyinghead** [comment]
Are you using HLE BIOS?

The only way I was able to produce some cracking was to use HLE BIOS. Because it reads all sectors requested by the game in one shot. In Soulcalibur I see read times of ~50 ms for 500 to 700 sectors at once, which is a lot.

When using real BIOS, max 32 sectors are read at once with read times ~3 ms each, which is fine.

**Comment by flyinghead** [comment]
There's nothing wrong about reporting a bug with HLE bios, quite the contrary in fact!

Anyway I tried again with the same configuration (RA, alsathread with 32 ms latency, vulkan per pixel, ubuntu 20) and there's no audio issue with Soulcalibur. I would suggest you look at your RA and system settings to find out what is causing this issue. I will fix the HLE BIOS issue but I don't see any with regular BIOS. I suspect you might have similar issues when flycast recompile dynarec codes, load textures or compile shaders. All these operations can sometimes take time but shouldn't be noticeable on a normally configured system.

**Comment by flyinghead** [comment]
I have fixed the issue with HLE BIOS. I also reduced by half the size of the gdrom buffer at 16 buffers.
This might help the issue you're seeing.

However I think the real issue here is in your hardware or settings. And you should probably relax the retroarch settings a bit to avoid running at the limit of what your system can deliver.

**Comment by vanfanel** [comment]
Ah! I hope to see this on the Master branch soon! :)

**Comment by vanfanel** [comment]
@flyinghead  Sorry to reopen this, but I am sometimes still hearing crackling/dropouts with 32/36ms@48KHz audio buffer with direct ALSA here, be it with CHD or CUE/BIN/GDI uncompressed images.
Please note that, unless you are on a system with real ALSA only (NO PulseAudio at all) you may have never really gotten the specified audio buffer!

Even copying the game image to /tmp (which is mounted in RAM) makes no difference, so image access speed is not the problem.

The setting that really alleviates those dropouts/cracklings with relatively small audio buffers is `Emulation Hacks -> GD-ROM Fast Loading`.  Activating that, there are ZERO audio problems and the sound buffer can be lowered as low as with other cores, but of course that's a hack and introduces some small problems in games.

So I believe there's something on the GD-ROM accurate access routines blocking more than it should.
Since I do my own builds and I can read C, can you point me to the code that introduces blocking for accurate GD-ROM access times simulation? I would like to experiment with it to help fixing this, as low-latency audio in emulation is important to me and I have helped fixing these problems in other cores.

Also, what about threaded GD-ROM access? PSX cores use threaded CD-ROM access to avoid this kind of problems.

**Comment by Nejsplitter** [comment]
Hi @vanfanel , fyi regarding the audio buffer/latency setting in retroarch, when dynamic rate control is enabled then retroarch will target 50% saturation of the audio buffer, which means 64ms latency becomes 32ms and 32ms latency becomes 16ms etc

**Comment by vanfanel** [comment]
> Hi @vanfanel , fyi regarding the audio buffer/latency setting in retroarch, when dynamic rate control is enabled then retroarch will target 50% saturation of the audio buffer, which means 64ms latency becomes 32ms and 32ms latency becomes 16ms etc

Yes, you're right, but 64ms as a setting is relatively the same buffer size with this core and with other cores, and I am only seeing this problem with this core.
